Added discord-app-commands.md (#733)
discord-python/site: pydis_site/apps/content/resources/guides/python-guides/discord-app-commands.md

---
title: Slash Commands with discord.py!
description: A simple guide to creating slash commands within discord.py!
---
# DISCORD.PY RESUMPTION CHANGES

---

> Upon resumption of the most popular discord API wrapper library for python, `discord.py`, while catching on to the latest features of the discord API, there have been numerous changes with addition of features to the library. Some additions to the library are -> Buttons support, Select Menus Support, Forms (AKA Modals), Slash Commands (AKA Application Commands) and a bunch of more handy features! All the changes can be found [here](https://discordpy.readthedocs.io/en/latest/migrating.html). Original discord.py gist regarding resumption can be found [here](https://gist.github.com/Rapptz/c4324f17a80c94776832430007ad40e6).

Reviewer: Let's try to use more markdown features so things don't feel too cluttered.
```suggestion
Upon resumption of the most popular discord API wrapper library for python, `discord.py`, while catching on to the latest features of the discord API, there have been numerous changes with addition of features to the library. Some additions to the library include: 

- Buttons support
- Select Menus support
- Forms (AKA Modals)
- Slash commands (AKA Application Commands)
...and a bunch more handy features!

All the changes can be found [here](https://discordpy.readthedocs.io/en/latest/migrating.html). Original discord.py gist regarding resumption can be found [here](https://gist.github.com/Rapptz/c4324f17a80c94776832430007ad40e6).
```

async def my_setup_hook() -> None:
    #perform tasks

client.setup_hook = my_setup_hook
```

# Basic Slash Command application using discord.py.

#### The `CommandTree` class resides within the `app_commands` of discord.py package.
---

## Slash Command Application with a Client

```python
import discord

class SlashClient(discord.Client):
    def __init__(self) -> None:
        super().__init__(intents=discord.Intents.default())
        self.tree = discord.app_commands.CommandTree(self)

    async def setup_hook(self) -> None:
        self.tree.copy_global_to(guild=discord.Object(id=12345678900987654))
        await self.tree.sync()

client = SlashClient()

@client.tree.command(name="ping", description="...")
async def _ping(interaction: discord.Interaction) -> None:
    await interaction.response.send_message("pong")

client.run("token")
```

__**EXPLANATION**__

- `import discord` imports the **discord.py** package.
- `class SlashClient(discord.Client)` is a class subclassing **Client**. Though there is no particular reason except readability to subclass the **Client** class, using the `Client.setup_hook = my_func` is equally valid.

> - Next up `super().__init__(...)` runs the `__init__` function of the **Client** class, this is equivalent to `discord.Client(...)`. Then, `self.tree = discord.app_commands.CommandTree(self)` creates a CommandTree which acts as the container for slash commaands.

Reviewer: We also want to explain why we're binding it to `self`
```suggestion
- Next up `super().__init__(...)` runs the `__init__` function of the **Client** class, this is equivalent to `discord.Client(...)`. Then, `self.tree = discord.app_commands.CommandTree(self)` creates a CommandTree which acts as the container for slash commands, and binds it to the `discord.Client` subclass instance, so wherever you have access to it, you will also have access to the command tree.
```

> - Then in the `setup_hook`, `self.tree.copy_global_to(...)` adds the slash command to the guild of which the ID is provided as a `discord.Object` object. Further up, `self.tree.sync()` updates the API with any changes to the slash commands.

Reviewer: Forgetting to sync is a common mistake
```suggestion
- Then in the `setup_hook`, `self.tree.copy_global_to(...)` adds the slash command to the guild of which the ID is provided as a `discord.Object` object. Further up, `self.tree.sync()` updates the API with any changes to the slash commands. **Without calling this method, your changes will only be saved locally and will NOT show up on Discord!**
```

- And the classic old `client.run("token")` is used to connect the client to the discord gateway.
- Note that the `send_message` is a method of the `InteractionResponse` class and `interaction.response` in this case is an instance of the `InteractionResponse` object. The `send_message` method will not function if the response is not sent within 3 seconds of command invocation. I will discuss about how to handle this issue later following the gist.

## Slash Command Application with a Bot

```python
import discord

class SlashBot(commands.Bot):
    def __init__(self) -> None:
        super().__init__(command_prefix=".", intents=discord.Intents.default())

    async def setup_hook(self) -> None:
        self.tree.copy_global_to(guild=discord.Object(id=12345678900987654))
        await self.tree.sync()

bot = SlashBot()

@bot.tree.command(name="ping", description="...")

> async def _ping(interaction: discord.Interaction) -> None:

Reviewer: There's really no need to prefix the `ping` command with an underscore, and I believe it only serves to confuse newcomers.
```suggestion
async def ping(interaction: discord.Interaction) -> None:
```

    await interaction.response.send_message("pong")

bot.run("token")
```

The above example shows a basic slash commands within discord.py using the Bot class.

__**EXPLANATION**__

Most of the explanation is the same as the prior example which featured `SlashClient` which was a subclass of **discord.Client**. Though some minor changes are discussed below.

- The `SlashBot` class now subclasses `discord.ext.commands.Bot` following the passing in of the required arguments to its `__init__` method.
- `discord.ext.commands.Bot` already consists of an instance of the `CommandTree` class which can be accessed using the `tree` property.

> 

Reviewer: Could we also explain the fact that we can now use both slash commands *and* regular commands?

Author: That's a separate topic and this guide only focuses on slash commands. I wouldn't be agreeing on that idea but I leave the rest to your judgement

# Slash Commands within a Cog!

---

A cog is a collection of commands, listeners, and optional state to help group commands together. More information on them can be found on the [Cogs](https://discordpy.readthedocs.io/en/latest/ext/commands/cogs.html#ext-commands-cogs) page.

## An Example to using cogs with discord.py for slash commands!

```python
import discord
from discord.ext import commands
from discord import app_commands

class MySlashCog(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(name="ping", description="...")

>     async def _ping(self, interaction: discord.Interaction):

Reviewer: Again, no need to prefix with an underscore
```suggestion
    async def ping(self, interaction: discord.Interaction):
```

An example to creating an error handler for slash commands is as follows.

```python
from discord.ext import commands
from discord import app_commands
import discord

bot = commands.Bot(command_prefix="!", intents=discord.Intents.default())
#sync commands

@bot.tree.command(name="ping")
app_commands.checks.cooldown(1, 30)
async def ping(interaction: discord.Interaction):
    await interaction.response.send_message("pong!")

async def on_tree_error(interaction: discord.Interaction, error: app_commands.AppCommandError):
    if isinstance(error, app_commands.CommandOnCooldown):
        return await interaction.response.send_message(f"Command is currently on cooldown! Try again in **{error.retry_after:.2f}** seconds!")

>     elif isinstance(..., ...):

Reviewer: Just to make the pattern extra clear
```suggestion
    elif isinstance(error, ...):
```

        ...

    else:
        raise error

bot.tree.on_error = on_tree_error

bot.run("token")
```

__**EXPLANATION**__

First we create a simple function named as `on_tree_error` here. To which the first two required arguments are passed, `Interaction` which is named as `interaction` here and `AppCommandError` which is named as `error` here. Then using simple functions and keywords, we make an error handler like above. Here I have used the `isinstance` function which takes in an object and a base class as the second argument, this function returns a bool value. After creating the error handler function, we set the function as the error handler for the slash commands. Here, `bot.tree.on_error = on_tree_error` overwrites the default `on_error` method of the **CommandTree** class with our custom error handler which has been named as `on_tree_error` here.

> 

Reviewer: Mention the use of `raise error`, as we don't want people to think it's useless and remove it, and not have any more errors

### Creating an error handler for a specific error!

```python
from discord.ext import commands
from discord import app_commands
import discord

bot = commands.Bot(command_prefix="!", intents=discord.Intents.default())
#sync commands

@bot.tree.command(name="ping")

> app_commands.checks.cooldown(1, 30)

Reviewer: ```suggestion
@app_commands.checks.cooldown(1, 30)
```

async def ping(interaction: discord.Interaction):
    await interaction.response.send_message("pong!")

@ping.error
async def ping_error(interaction: discord.Interaction, error: app_commands.AppCommandError):
    if isinstance(error, app_commands.CommandOnCooldown):
        return await interaction.response.send_message(f"Command is currently on cooldown! Try again in **{error.retry_after:.2f}** seconds!")

>     elif isinstance(..., ...):

Reviewer: ```suggestion
    elif isinstance(error, ...):
```
